metriks/metriks.go

```diff
@@ -0,0 +1,124 @@
+package metriks
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/armon/go-metrics"
+	"github.com/armon/go-metrics/datadog"
+)
+
+const timerGranularity = time.Millisecond
+
+// Init will initialize the internal metrics system and build a Datadog statsd sink
+func Init(serviceName string, conf Config) error {
+	sink, err := datadog.NewDogStatsdSink(statsdAddr(conf), conf.Name)
+	if err != nil {
+		return err
+	}
+
+	var tags []string
+	for k, v := range conf.Tags {
+		tags = append(tags, fmt.Sprintf("%s:%s", k, v))
+	}
+
+	sink.SetTags(tags)
+
+	c := metrics.DefaultConfig(serviceName)
+	c.TimerGranularity = timerGranularity
+
+	if _, err := metrics.NewGlobal(c, sink); err != nil {
```

Comment:
Setting a global metrics object won't allow us to eventually use different sinks in the same app. I.e., we may want to use Datadog for one mod but Prometheus for another one. I want also to emphasize the fact that this could cause problems in tests around metrics and collisions between tests in case different sinks need to be used.

Reply:
So metrics are one of those cross-cutting concerns that I feel you really want to be able to use anywhere in a code base easily, similar to logging. I think there's a lot of value in keeping support for a global metriks scope, because it means you can immediately instrument any piece of code, no matter how deep in an abstraction layer. From past experience, I've found that if you add barriers to instrumentation, even minor, it can tend to lead to skipping it in places you really would prefer to have it. You don't want to have to balance between refactoring multiple layers of abstraction to pass a metrics object so that you can quickly instrument an error count, for example.

However, the global model works best when it's simply internal service / application code and we've all agreed on a shared instrumentation library. If we were building shared libraries that we wanted to open source and allow others to use in applications outside Netlify, then global metrics registry approach wouldn't work and we'd want a configurable metrics dependency. For this, I think we could always take this current interface and make it a proper exported interface that any open source library could take as an interface and invoke directly on. We probably want to wait until we've settled on one across projects though.

> we may want to use Datadog for one mod but Prometheus for another one

I'm hoping we never want to split metrics for a service tier between different metrics systems. :-D That being said, if we need to we could build an interface from this that certain modules can take a metrics object, instead of using the global. I think that'd be best for code shared outside of our org.

Reply:
I agree on the global settings. I think that it can be very nice to not have to pass the metrics connection all the way down. It is nice in logs b/c we're able to maintain state (e.g. the request id) but that isn't really needed in the case of metrics. 

> we may want to use Datadog for one mod but Prometheus for another one

No please 😭. I think that the `InitWithSink` is a good idea, we can test with feature flags for percentages of the service, not running two sinks in the same service. Also, we _just_ signed a contract with DataDog, this will be a moot point for at least another year.